content/advent-2018/go-in-the-browser.md

```diff
@@ -0,0 +1,175 @@
+-rwxrwxr-x 1 agniva agniva 6721905 Sep 24 12:28 serve
+-rw-rw-r-- 1 agniva agniva      76 Dec  6 12:08 test.go
+-rwxrwxr-x 1 agniva agniva 2425246 Dec  6 12:09 test.wasm
+-rw-r--r-- 1 agniva agniva   11905 Dec  6 12:16 wasm_exec.js
+```
+
+`serve` is a simple Go binary to serve files in the current directory. But just about any web server will do.
+
+Once we run this, and open our browser. We see a `Run` button, clicking which, will execute our application. Then we click it and check the console -
+
+![hellowasm](/postimages/advent-2018/go-in-the-browser/hellowasm.png)
+
+Neat ! We just wrote a program in Go and ran it inside the browser.
+
+So far so good. But this was a simple hello world program. A real world webapp needs to interact with the DOM. We need to respond to button click events, take input data from text boxes, and send data back to the DOM. Now we will build a minimal image editor in the browser which will use all of these capabilities.
+
+## DOM API
+
+But first, to interact with the browser from Go code, we need a DOM API. We have the `syscall/js` library to help us out with that. It is a very basic, but nevertheless powerful form of a DOM API, on top of which we can build our app. Let us very quickly see some of its capabilities before we move over to making our app.
+
+**Callbacks**
+
+To respond to DOM events, we declare callbacks and hook them up with events like this -
+
+```
+import “syscall/js”
+
+// Declare callback
+cb := js.NewEventCallback(js.PreventDefault, func(ev js.Value) {
+	// handle event
+})
+
+
+// Hook it up with a DOM event
+js.Global().Get("document").
+	Call("getElementById", "myBtn").
+	Call("addEventListener", "click", cb)
+
+
+// Call cb.Release() on your way out.
+```
+
+**Updating the DOM**
+
+To update the DOM from inside Go, we can do-
+
+```
+import “syscall/js”
+
+js.Global().Get("document").
+		Call("getElementById", "myTextBox").
+		Set("value", "hello wasm")
+```
+
+You can even call JS functions and manipulate native native JS Objects like `FileReader` or `Canvas`. Feel free to check out the [syscall/js](https://golang.org/pkg/syscall/js/) documentation for further details.
+
+Ok, now on with building our app !
+
+## A proper webapp
+
+We will build a small app which will take an input image, then perform some manipulations on the image like brightness, contrast, hue, saturation, and finally send the output image back to the browser. There will be sliders for each of these effects, which the user can change and see the target image change in real time.
+
+First, we need to get the input image from the browser to our Go code, so that we can work on it. To efficiently do this, we need to resort to some `unsafe` tricks, the details of which I will skip here. Once we have the image, it is fully in our control and we are free to do whatever with it. Below is a brief snippet from the image loader callback -
+
+```go
+s.onImgLoadCb = js.NewCallback(func(args []js.Value) {
```

Comment:
Could I suggest to get rid of `s` variable in examples? I don't think it adds any value to what you're describing in the text, but it makes snippets more complicated.

How about you update the examples as the following:

```go
onImgLoadCb := js.NewCallback(func(args []js.Value) {  
    ···
})

js.Global().Set("loadImage", onImgLoadCb)
```

For me, as for a person who isn't familiar with the API of `syscall/js`, but who's more or less familiar with DOM API, this helps to create some meaningful connections between things you're showing.

Reply:
I actually wanted to show the code inside the callback to show how we get the image inside Go. I will try to simplify it.

Reply:
Sure. By `···` I meant the code you have now but without `s` receiver.